src/community/web-resources/src/main/java/org/geoserver/web/resources/PageResourceBrowser.java

```diff
@@ -0,0 +1,516 @@
+                    protected Component getContents(String id) {
+                        //pick a non-existing resource name (can be changed by user)
+                        String dest = "/" + Paths.path(treeView.getSelectedNode().getObject().path(), "new.txt");
+                        for (int i = 1; Resources.exists(store().get(dest)); i++) {
+                            dest = "/" + Paths.path(treeView.getSelectedNode().getObject().path(), "new." + i + ".txt");
+                        }
+                        editPanel = new PanelEdit(id, dest, true, "");
+                        return editPanel;
+                    }
+
+                    @Override
+                    protected boolean onSubmit(AjaxRequestTarget target, Component contents) {
+                        editPanel.getFeedbackMessages().clear();
+                        Resource dest = store().get(editPanel.getResource());
+                        if (Resources.exists(dest)) {
+                            editPanel.error(new ParamResourceModel("resourceExists", getPage()).getString());
+                        } else {
+                            try (OutputStream os = dest.out()) {
+                                String newContents = editPanel.getContents();
+                                if (newContents != null) {
+                                    os.write(newContents.getBytes());
+                                }
+                                //select newly created node
+                                treeView.setSelectedNode(new ResourceNode(dest, expandedStates), target);
+                                return true;
+                            } catch (IOException | IllegalStateException e) {
+                                error(e.getMessage());
+                            }
+                        }
+                        target.add(editPanel.getFeedbackPanel());
+                        return false;
+                    }
+                });
+
+            }
+            
+        };
+        
+        //download button
+        final Link<Void> btnDownload = new Link<Void>("download") {
+            private static final long serialVersionUID = 2746429086122117005L;
+
+            @Override
+            public void onClick() {
+                Resource res = treeView.getSelectedNode().getObject();
+                getRequestCycle().scheduleRequestHandlerAfterCurrent(
+                        new ResourceStreamRequestHandler(new WicketResourceAdaptor(res))
+                            .setFileName(res.name())
+                            .setContentDisposition(ContentDisposition.ATTACHMENT));
+            }
+            
+        };
+        
+        //edit button
+        final AjaxLink<Void> btnEdit = new AjaxLink<Void>("edit") {
+            private static final long serialVersionUID = 6690936054046040647L;
+
+            @Override
+            public void onClick(AjaxRequestTarget target) {
+                dialog.setInitialHeight(500);
```

Comment:
The height depends on the content I am going to put in the modal window, which is different in each case.